### DISCUSSION TITLE
`pywraplp.Solver` deep copy

### DISCUSSION THREAD

**Discussion by SanPen** [google/or-tools#3960] [Linear Solver questions]
**What language and solver does this apply to?**
Python

**Describe the problem you are trying to solve.**

I want to copy a `Solver` object and in the copy, add a variable to each constraint.

These two features apply:
- Deep copy of the Solver object.
- Access to the variables used by a constraint and modify a constraint

```python
from ortools.linear_solver import pywraplp

def create_constraints(model):
    x = model.NumVar(0, 1, 'x')
    y = model.NumVar(0, 1, 'y')
    constraint = model.Constraint(-pywraplp.Solver.infinity(), 1)
    constraint.SetCoefficient(x, 1)
    constraint.SetCoefficient(y, 1)
    return constraint

# Create the first solver and add constraints
solver1 = pywraplp.Solver.CreateSolver('SCIP')
constraint1 = create_constraints(solver1)

# Create the second solver
new_solver = pywraplp.Solver.CreateSolver('SCIP')

# copy vars... kinda ok
for var in solver.variables():
    new_solver.NumVar(lb=var.Lb(), ub=var.Ub(), name=var.name())

# I want to copy the constraints, but adding a new var to each one
slacks = list()
for i, cst in enumerate(solver.constraints()):
    sl = new_solver.NumVar(0, 1e20, 'Slack{}'.format(i))
    new_solver.Add(cst + sl)  # this line fails
    slacks.append(sl)

new_solver.Minimize(new_solver.Sum(slacks))
```

**Describe the solution you'd like**

I'd like something like `solver.copy()` to copy the solver.

I'd like an easy way to add variables to an existing constraint (if this does not exist already, which I haven't been able to accomplish)

**Describe alternatives you've considered**

Not sure I have alternatives for this

**Additional context**

I want to add MIP model debugging by the process described in the code

asked on Stack Overflow: https://stackoverflow.com/questions/77348474/ortools-mip-solver-copy-pywraplp-solver-copy

**Comment by lperron** [comment]
Just answered on stack overflow.
